backend/tournesol/lib/suggestions/strategies/tocompare/classic.py

```diff
@@ -0,0 +1,254 @@
+        """
+        extra_sample1 = 0
+        extra_sample2 = 0
+
+        free_slots_in_pool1 = pool1.sample_size - len(pool1.ids[: pool1.sample_size])
+        free_slots_in_pool2 = pool2.sample_size - len(pool2.ids[: pool2.sample_size])
+
+        # If the pool 1 contains fewer ids than expected, try to pick more
+        # ids from the pools 2.
+        if free_slots_in_pool1 > 0:
+            extra_sample2 = free_slots_in_pool1
+
+        # If the pool 2 contains fewer ids than expected, try to pick more
+        # ids from the pool 1.
+        if free_slots_in_pool2 > 0:
+            extra_sample1 = free_slots_in_pool2
+
+        sample1 = pool1.ids[: pool1.sample_size + extra_sample1]
+        sample2 = pool2.ids[: pool2.sample_size + extra_sample2]
+        sample3 = pool3.ids[: pool3.sample_size]
+
+        free_slots = self.max_suggestions - len(sample1) - len(sample2) - len(sample3)
+
+        if free_slots > 0:
+            extra_sample3 = free_slots // 2
+
+            if free_slots % 2 == 1:
+                extra_sample3 += 1
+
+            sample3 = pool3.ids[: pool3.sample_size + extra_sample3]
+
+        return sample1 + sample2 + sample3
+
+    def get_serializer_class(self):
+        return EntityToCompare
+
+    def get_results(self):
+        return self.get_results_for_user_intermediate()
+
+    def get_results_for_user_new(self):
+        raise NotImplementedError
+
+    def get_results_for_user_intermediate(self):
+        poll = self.poll
+
+        pool1 = self._ids_from_pool_compared()
+        pool2 = self._ids_from_pool_rate_later(pool1)
+        pool3 = self._ids_from_pool_reco_last_month(pool1 + pool2)
+
+        sample1_size = len(pool1[: self.sample_size_compared])
+        sample2_size = len(pool2[: self.sample_size_rate_later])
+        sample3_size = len(pool3[: self.sample_size_reco_last_month])
+
+        if sample1_size + sample2_size + sample3_size >= self.max_suggestions:
+            return Entity.objects.filter(
+                id__in=pool1[: self.sample_size_compared]
+                + pool2[: self.sample_size_rate_later]
+                + pool3[: self.sample_size_reco_last_month]
+            ).with_prefetched_poll_ratings(poll_name=poll.name)
+
+        # Allow the empty slots from the pool "compared" to be filled by the
+        # items of the pool "rate-later" and vice-versa.
+        results = self._consolidate_results(
+            IdPool(pool1, self.sample_size_compared),
+            IdPool(pool2, self.sample_size_rate_later),
+            IdPool(pool3, self.sample_size_reco_last_month),
+        )
+
+        free_slots = self.max_suggestions - len(results)
+
+        if free_slots > 0:
+            last_resort = self._ids_from_pool_reco_all_time(results)
+            results += last_resort[:free_slots]
+
+        return Entity.objects.filter(id__in=results).with_prefetched_poll_ratings(
+            poll_name=poll.name
+        )
```

Comment:
Should we shuffle the list of entities when using `id__in`? The order of the results returned by the database is not deterministic, but probably not very random either.

Reply:
Exactly, it's better to shuffle the results.

Reply:
Fixed by https://github.com/tournesol-app/tournesol/pull/1809/commits/733e0ac5f0536c74fed0c8fc0dd979ad18e59f87